test/_detail_number.cpp

```diff
@@ -36,12 +38,111 @@ operator==(
     default:
         break;
     }
-    return 
+    return
         std::signbit(lhs.d) ==
             std::signbit(rhs.d) &&
         lhs.d == rhs.d;
 }
 
+struct double_diagnoser
+{
+
+    static void emit_hex(std::ostream& os, double d)
+    {
+        std::uint64_t binary;
+        static_assert(sizeof(binary) == sizeof(d), "");
+
+        std::memcpy(&binary, &d, sizeof(d));
+        auto oldflags = os.flags();
+
+        try
+        {
+            os << std::hex << std::setw(16) << std::setfill('0') << binary;
+            os.flags(oldflags);
+        }
+        catch(...)
+        {
+            os.flags(oldflags);
+        }
+    }
+    static void emit_pow2(std::ostream& os, double d)
+    {
+        auto oldflags = os.flags();
+        try
+        {
+            int exponent = 0;
+            auto mantissa = std::frexp(d, &exponent);
+            os << std::fixed << mantissa;
+            os << " *2^ ";
+            os << exponent;
+            os.flags(oldflags);
+        }
+        catch(...)
+        {
+            os.flags(oldflags);
+            throw;
+        }
+    }
+
+    static void emit_scientific(std::ostream& os, double d)
+    {
+        auto oldflags = os.flags();
+        try
+        {
+            constexpr auto digits = std::numeric_limits<double>::max_digits10;
```

Comment:
what is `constexpr` buying us here and why even have `digits` at all? i.e. why not
```
std::setprecision(std::numeric_limits<double>::max_digits10)
```
?

When we name a thing we elevate its importance, we should not elevate things needlessly.

Reply:
The reason I do this:
Spelling out concepts one per line is to me the same as your preference for formatting code in terms of one piece of entropy per line.
Naming a temporary variable like this is a zero-cost abstraction. `constexpr` is a signal that this is a 'number concept' as opposed to an actual value object.

Nevertheless, your code, your rules. Will change.